src/surface/flight_control/flight_control/mavlink_control_node.py

```diff
@@ -0,0 +1,468 @@
+
+        if not self.vehicle_state.pi_connected:
+            self.vehicle_state.pi_connected = True
+            self.publish_state(self.vehicle_state)
+            self.get_logger().info('Pi connected')
+
+    def poll_subscribers(self) -> None:
+        # Whenever a node subscribes to vehicle state updates, send the current state
+        subscriber_count = self.state_publisher.get_subscription_count()
+        if subscriber_count > self.last_state_subscriber_count:
+            self.publish_state(self.vehicle_state)
+
+        self.last_state_subscriber_count = subscriber_count
+
+    def handle_pygame_events(self) -> None:
+        for event in pygame.event.get():
+            if event.type == pygame.JOYDEVICEADDED:
+                new_joy = joystick.Joystick(event.device_index)
+                self.joysticks[new_joy.get_instance_id()] = new_joy
+                if not self.current_joystick_id:
+                    self.current_joystick_id = new_joy.get_instance_id()
+                self.get_logger().info(f'Joystick {new_joy.get_instance_id()} connected')
+
+            elif event.type == pygame.JOYDEVICEREMOVED:
+                del self.joysticks[event.instance_id]
+                if self.current_joystick_id == event.instance_id:
+                    if self.joysticks:
+                        # Arbitrarily choose a new joystick to become the current joystick
+                        # In practice this is the first element of list(self.joystick.keys())
+                        self.current_joystick_id = next(iter(self.joysticks))
+                    else:
+                        self.current_joystick_id = None
+                self.get_logger().info(f'Joystick {event.instance_id} disconnected')
+
+    def poll_joystick(self) -> None:
+        """Read the current state of the joystick and send a mavlink message."""
+        self.handle_pygame_events()
+
+        if self.current_joystick_id is not None:
+            joy = self.joysticks[self.current_joystick_id]
+
+            axes = [joy.get_axis(ax) for ax in range(joy.get_numaxes())]
+            buttons = [joy.get_button(btn) for btn in range(joy.get_numbuttons())]
+
+            # Represent dpad movements as button presses
+            for hat_idx in range(joy.get_numhats()):
+                for hat_axis in joy.get_hat(hat_idx):
+                    if hat_axis > 0:
+                        buttons += [False, True]
+                    elif hat_axis < 0:
+                        buttons += [True, False]
+                    else:
+                        buttons += [False, False]
+
+            self.controller_callback(JoystickState(axes=axes, buttons=buttons))
+
+
+def main() -> None:
+    rclpy.init()
+    manual_control = MavlinkManualControlNode()
+    executor = MultiThreadedExecutor()
+
+    last_loop_start = time.time()
+    last_mav_poll = time.time()
+    last_sub_poll = time.time()
+
+    loop_period = 1 / JOYSTICK_POLL_RATE
```

Comment:
Mayve can do something nice with manual_control.create_rate instead of doing by hand?

Reply:
I want to control this loop as tightly as possible, idk if I trust ROS with it